File: GLLara/GLLMeshController.m
//
//  GLLMeshController.m
//  GLLara
//
//  Created by Torsten Kammer on 01.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GLLMeshController.h"

#import "GLLItemMesh.h"

@implementation GLLMeshController

- (id)initWithMesh:(GLLItemMesh *)mesh parentController:(id)parentController;
{
	if (!(self = [super init])) return nil;
	
	self.mesh = mesh;
	self.parentController = parentController;
	
	return self;
}

- (id)representedObject
{
	return self.mesh;
}

#pragma mark - Outline View Data Source

- (id)outlineView:(NSOutlineView *)outlineView child:(NSInteger)index ofItem:(id)item
{
	return nil;
}

- (id)outlineView:(NSOutlineView *)outlineView objectValueForTableColumn:(NSTableColumn *)tableColumn byItem:(id)item
{
	return self.mesh.displayName;
}

- (void)outlineView:(NSOutlineView *)outlineView setObjectValue:(id)object forTableColumn:(NSTableColumn *)tableColumn byItem:(id)item
{
	self.mesh.displayName = object;
}

- (BOOL)outlineView:(NSOutlineView *)outlineView isItemExpandable:(id)item
{
	return NO;
}

- (NSInteger)outlineView:(NSOutlineView *)outlineView numberOfChildrenOfItem:(id)item
{
	return 0;
}

#pragma mark - Outline View Delegate

- (BOOL)outlineView:(NSOutlineView *)outlineView shouldEditTableColumn:(NSTableColumn *)tableColumn item:(id)item
{
	return YES;
}

@end

File: GLLara/GLLMeshController.swift
//
//  GLLMeshController.m
//  GLLara
//
//  Created by Torsten Kammer on 01.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

import Cocoa

/**
 * @abstract Source list controller for a mesh.
 */
@objc class GLLMeshController : NSObject, NSOutlineViewDataSource, NSOutlineViewDelegate {
    init(mesh: GLLItemMesh, parentController: AnyObject) {
        self.mesh = mesh
        self.parentController = parentController
    }
    
    let mesh: GLLItemMesh
    @objc weak var parentController: AnyObject?
    @objc var representedObject: GLLItemMesh {
        return mesh
    }
    
    func outlineView(_ outlineView: NSOutlineView, objectValueFor tableColumn: NSTableColumn?, byItem item: Any?) -> Any? {
        return mesh.displayName
    }
    
    func outlineView(_ outlineView: NSOutlineView, setObjectValue object: Any?, for tableColumn: NSTableColumn?, byItem item: Any?) {
        if let newName = object as? String {
            mesh.displayName = newName
        }
    }
    
    func outlineView(_ outlineView: NSOutlineView, isItemExpandable item: Any) -> Bool {
        return false
    }
    
    func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
        return 0
    }
    
    func outlineView(_ outlineView: NSOutlineView, shouldEdit tableColumn: NSTableColumn?, item: Any) -> Bool {
        return true
    }

}
